### PR TITLE
Storefront API docs

### PR DESCRIPTION
**Pull request by mattbrailsford**
## Description

This PR adds documentation for the new Umbraco Commerce Storefront API. 

## Type of suggestion

* [ ] Typo/grammar fix
* [ ] Updated outdated content
* [x] New content
* [ ] Updates related to a new version
* [ ] Other

## Product & version (if relevant)
Umbraco Commerce v12.1.0

### CODE REVIEW

File: 12/umbraco-commerce/reference/storefront-api/README.md
Comment (sofietoft) [inline 12/umbraco-commerce/reference/storefront-api/README.md:7]: Shortened a sentence.

```suggestion
The Storefront API delivers headless capabilities built directly into Umbraco Commerce. It allows you to retrieve and manage orders and other store-related entities in a JSON format. It lets you connect with them in different channels, using your preferred technology stack. This feature preserves the friendly editing experience of Umbraco, while also ensuring performant cart management facilities in a headless fashion. With its different extension points, you can tailor this API to fit a broad range of requirements.
```

File: 12/umbraco-commerce/reference/storefront-api/README.md
Comment (sofietoft) [inline 12/umbraco-commerce/reference/storefront-api/README.md:11]:
```suggestion
As the Storefront API works alongside the Content Delivery API you must first have the [Content Delivery API setup and enabled](https://docs.umbraco.com/umbraco-cms/reference/content-delivery-api#register-the-content-delivery-api-dependencies).

When the Content Delivery API is enabled, you will need to explicitly opt-in to the Storefront API. Below you will find the steps you need to take in order to configure it for your Umbraco project.
```

File: 12/umbraco-commerce/reference/storefront-api/README.md
Comment (sofietoft) [inline 12/umbraco-commerce/reference/storefront-api/README.md:56]:
```suggestion
In order to work with the Storefront API many of the endpoints require authorization. The authorization is implemented by means of an API Key that must be passed in the header of these requests. The API Key is defined as an additional app setting and can be any value you decide:
```

File: 12/umbraco-commerce/reference/storefront-api/README.md
Comment (sofietoft) [inline 12/umbraco-commerce/reference/storefront-api/README.md:81]:
```suggestion
When working with Umbraco Commerce's C# Api, Umbraco Commerce will normally keep track of a series of items for you. This could be the current Order ID or the current Language, which it tracks in a cookie. When working in a headless manner however we can no longer rely on cookies for this behavior. It becomes the implementor's responsibility to remember these items and pass them as Headers to the endpoints that need to use them for context.
```

File: 12/umbraco-commerce/reference/storefront-api/README.md
Comment (sofietoft) [inline 12/umbraco-commerce/reference/storefront-api/README.md:94]:
```suggestion
* `Store` - The ID or Alias of the store the given operation is being performed against.
* `Current-Order` - The ID of any current "in-progress" order.
* `Customer-Reference` - A unique reference for the current customer.
* `Accept-Language` - The ISO Culture Code of the current front-end language.
* `Currency` - The ID or ISO Code of the current currency. If not supplied, will fall back to the default currency of the default country defined on the store.
* `Tax-Class` - The ID or Alias of the default tax class. If not supplied, will fall back to the default defined on the store.
* `Billing-Country` - The ID or ISO Code of the default billing country. If not supplied, falls back to the default country defined on the store.
* `Billing-Region` - The ID or ISO Code of the default billing region. 
* `Shipping-Country` - The ID or ISO Code of the default shipping country. If not supplied, falls back to either the supplied billing country or the default country defined on the store.
* `Shipping-Region` - The ID or ISO Code of the default billing region.
```

File: 12/umbraco-commerce/reference/storefront-api/README.md
Comment (sofietoft) [inline 12/umbraco-commerce/reference/storefront-api/README.md:104]:
```suggestion
When resources are connected, like an Order holding a reference to its Currency, the Storefront API will return those connected resources as "Reference" objects. This object will usually only contain the resource ID and Alias/Code. 
```

File: 12/umbraco-commerce/reference/storefront-api/README.md
Comment (sofietoft) [inline 12/umbraco-commerce/reference/storefront-api/README.md:121]:
```suggestion
Should you wish to retrieve the connected resource, you can perform a `get` operation against that resource's own endpoint. However, this could result in multiple wasteful network requests. Instead, you can pass an `expand` query parameter with a path to properties you wish to expand. For the targeted properties, this query will then return the full resource object instead of the reference entity.
```

File: 12/umbraco-commerce/reference/storefront-api/README.md
Comment (sofietoft) [inline 12/umbraco-commerce/reference/storefront-api/README.md:157]:
```suggestion
The `expand` query parameter can accept a comma-separated list of property keys to expand for properties at the same level. It can also expand nested objects using a `[...]` syntax. In the following example, we retrieve a Country entity. We will expand its `defaultCurrency` and `defaultPaymentMethod` whilst at the same time expand the `allowedCountries.country` properties within the `defaultCurrency`.
```

File: 12/umbraco-commerce/reference/storefront-api/README.md
Comment (sofietoft) [inline 12/umbraco-commerce/reference/storefront-api/README.md:224]:
```suggestion
Prices in Umbraco Commerce can contain a lot of metadata, such as listing applied discounts/adjustments and all values including or excluding those discounts. These are useful when displaying a full order breakdown, but when only needing to present an order's total value, are not necessary. By default, the Storefront API will only return the final `value` property of a price object. Should you wish to receive the full pricing breakdown you can pass the `$prices` shortcut key to expand all price values to include metadata.
```

File: 12/umbraco-commerce/reference/storefront-api/README.md
Comment (sofietoft) [inline 12/umbraco-commerce/reference/storefront-api/README.md:309]:
```suggestion
A common pitfall of REST APIs is the problem of over-fetching, which is where an endpoint returns more information that you need. The Storefront API supports the passing of `fields` which allows you to define exactly which properties of an object you wish to return. This will reduce the payload size.
```

File: 12/umbraco-commerce/reference/storefront-api/README.md
Comment (sofietoft) [inline 12/umbraco-commerce/reference/storefront-api/README.md:463]:
```suggestion
When using the `fields` query parameter to limit fields returned, `id` properties will always be included.
```

File: 12/umbraco-commerce/reference/storefront-api/README.md
Comment (sofietoft) [inline 12/umbraco-commerce/reference/storefront-api/README.md:611]:
```suggestion
* `attributes` will contain a list of "in use" attributes which means there is at least one variant content item that makes use of that attribute. These should be used to build the attribute options UI.
* `items` returns a list of variant items. By default, this will return the attribute combinations, and whether it is the default combination but its content value will be empty. The content value can be populated by expanding the variants property through the Delivery API, however, it's important to know this could return a lot of data and be intensive. Instead, it is preferred to return the non-expanded value and use the `variantContentUrl` to fetch individual content items as they are requested. The `items` collection should also be used to check if a combination exists as whilst the root level `attributes` collection contains all in-use attributes, it doesn't mean every possible combination of those attributes exists so you can use the `items` collection to validate a combination.
* `variantContentUrl` as the URL to a specialized Delivery API route that can return a single variant item content based on a passed-in attribute combination. 
```

File: 12/umbraco-commerce/reference/storefront-api/endpoints/README.md
Comment (sofietoft) [inline 12/umbraco-commerce/reference/storefront-api/endpoints/README.md:1]: I'm wondering whether this landing page is necessary?
Wouldn't it be a little cleaner if the endpoints where direct subpages to the Storefront API README article?

File: 12/umbraco-commerce/reference/storefront-api/endpoints/checkout.md
Comment (sofietoft) [inline 12/umbraco-commerce/reference/storefront-api/endpoints/checkout.md:27]:
```suggestion
# Storefront Checkout API Endpoints

The checkout endpoints provide ways of performing a checkout process against an Order. The Storefront API supports two ways of checking out an order, one using hosted checkout pages, and a more advanced option for inline payment processing.

## Hosted

With the hosted checkout flow it is required that before redirecting to the payment gateway a checkout token should be generated. This token is passed to the pay endpoint to ensure that only the given order can be processed in response to the checkout request. The pay endpoint should then be launched in a WebView with this token which will then redirect to the given Orders payment gateway for payment capture. To determine the outcome of the payment developers should monitor the WebViews URL which will be redirected to the same pay endpoint URL with either a `/completed`, `/canceled` or `/errored` suffix.

{% swagger src=".gitbook/assets/storefront_swagger.json" path="/umbraco/commerce/storefront/api/v1/checkout/{orderId}/token" method="get" %}
[storefront_swagger.json](.gitbook/assets/storefront_swagger.json)
{% endswagger %}

{% swagger src=".gitbook/assets/storefront_swagger.json" path="/umbraco/commerce/storefront/api/v1/checkout/{orderId}/pay/{token}" method="get" %}
[storefront_swagger.json](.gitbook/assets/storefront_swagger.json)
{% endswagger %}

## Inline

With the inline checkout flow, it is left to the implementing developer to perform payment capture. Before a capture can begin, the order should be initialized using the `initialize` endpoint which prepares the Order for capture. To aid developers, the `initialize` endpoint will return the settings of the given Orders selected payment method, along with details of any expected metadata needed by the payment provider. Developers can then use this data to perform an inline capture operation and then call the `confirm` endpoint when the capture is successful, passing back any metadata captured.

{% swagger src=".gitbook/assets/storefront_swagger.json" path="/umbraco/commerce/storefront/api/v1/checkout/{orderId}/initialize" method="get" %}
[storefront_swagger.json](.gitbook/assets/storefront_swagger.json)
{% endswagger %}

{% swagger src=".gitbook/assets/storefront_swagger.json" path="/umbraco/commerce/storefront/api/v1/checkout/{orderId}/confirm" method="post" %}
[storefront_swagger.json](.gitbook/assets/storefront_swagger.json)
{% endswagger %}
```

**Review by sofietoft — CHANGES_REQUESTED** [review]
I've added some comments and suggestion here and there.
Please take a look to help ensure I didn't remove any of the meaning when shortening the long sentences.

Also, consider whether the Endpoint README.md file is necessary. Seems a little redundant to me.
All the endpoint articles are also missing from the SUMMARY file, but I didn't want to add them, until discussing this part about the landing page first. Let me know your thoughts!

I also noticed that the `product.md`  file is empty. Is that on purpose?
Should it be taken out of the list, if there's nothing in it anyways?

File: 12/umbraco-commerce/reference/storefront-api/endpoints/README.md
Comment (mattbrailsford) [reply 12/umbraco-commerce/reference/storefront-api/endpoints/README.md:1]: I was thinking it could be handy to just have a page to be able to point to for all the endpoints without having to scroll down a long main page. Also, linked to the SUMMARY file I wondered if it kept that less cluttered, but you can use your judgment on that 👍

**Comment by mattbrailsford** [comment]
Awesome stuff, thanks @sofietoft 

I'm happy with all the suggestions and have comitted them to the PR. There should indeed be a `product.md` file but i must have missed filling it in. I'll get that populated shortly.

Other than that, I think it sounds like it's just the decision around the endpoints page + side nav.

**Comment by mattbrailsford** [comment]
Product endpoints added and fixed another spelling issue too 👍

File: 12/umbraco-commerce/reference/storefront-api/endpoints/README.md
Comment (sofietoft) [reply 12/umbraco-commerce/reference/storefront-api/endpoints/README.md:1]: Ah! That's a good point. Let's keep the landing page there.

But all the articles need to be added to the SUMMARY file though.
Otherwise, they will not be part of the site on GitBook I'm afraid 😅

Comment (mattbrailsford) [reply 12/umbraco-commerce/reference/storefront-api/endpoints/README.md:1]: Do you want me to add them? Or you OK adding them in the same order as defined in the endpoints page?

Comment (sofietoft) [reply 12/umbraco-commerce/reference/storefront-api/endpoints/README.md:1]: I'm on it 💪

File: 12/umbraco-commerce/SUMMARY.md
Comment (sofietoft) [inline 12/umbraco-commerce/SUMMARY.md:70]: @mattbrailsford - I've updated the SUMMARY file now.

Another "funny" think with GitBook is that it wants to sync the link titles from the SUMMARY file with the titles of the articles they are referencing.
Currently, you've given the endpoint articles titles like `Storefront xxx API Endpoints`, but I've added only the `xxx` part to the SUMMARY file 😅 

I think that the entire title you've added is a little too long for the navigation, and would make it look a little cluttered.
How about we shorten the titles to something like `xxx Endpoints` instead?
We're already in the context of the Storefront API docs, so to me, it would be OK to leave those out.
What do you think?

Comment (mattbrailsford) [reply 12/umbraco-commerce/SUMMARY.md:70]: Yea, I'm cool with that 👍

Comment (mattbrailsford) [reply 12/umbraco-commerce/SUMMARY.md:70]: I never felt happy with them anyway to be honest

File: 12/umbraco-commerce/reference/storefront-api/endpoints/README.md
Comment (mattbrailsford) [reply 12/umbraco-commerce/reference/storefront-api/endpoints/README.md:1]: Legend 🙌

**Comment by sofietoft** [comment]
It looks like we're running into a similar issue, like we did on Andys forms PR, where the Swagger files aren't references quite right: https://github.com/umbraco/UmbracoDocs/pull/5308#issuecomment-1643978165 

The file is places correctly it looks like, however, the references to the files are not correct.
The links should be related to the folders that the file is in, so instead of
`src=".gitbook/assets/storefront_swagger.json"` and `[storefront_swagger.json](.gitbook/assets/storefront_swagger.json)`
it should be
`src="../../.gitbook/assets/storefront_swagger.json"` and `[storefront_swagger.json](../../.gitbook/assets/storefront_swagger.json)`

Does it make sense? 😄

**Comment by mattbrailsford** [comment]
Ahh yes it does.

It was a tricky one to work out as annoyingly GitBooks don't have any docs on how to do that so last week I setup a dummy GitBook account to set it up from the UI and try and figure out what it was doing in the markdown files. I guess I must have assumed the `.gitbook` folder was more of a global reference than needing to traverse back to it, but that makes total sense.

**Comment by mattbrailsford** [comment]
I'll go sort this and the headings issue now

**Comment by mattbrailsford** [comment]
All done, is that looking better @sofietoft ?

**Comment by sofietoft** [comment]
> All done, is that looking better @sofietoft ?

I took the liberty to make some adjustments on your fork, as it didn't look like the references were quite right still.
Hopefully they'll take this time 🤞

**Comment by sofietoft** [comment]
And it's working 🙌 😎 

And now we know how the Swagger API integration works - brilliant 😄 

Let me know when this should be merged!

**Comment by mattbrailsford** [comment]
🤦‍♂️ We'll get there eventually 😁

**Comment by alina-tincas** [comment]
Hi @mattbrailsford! 👋 

I was wondering if you know when this one should be merged? 😊

**Comment by mattbrailsford** [comment]
Hi @alina-tincas 

We're looking ideally to prep an RC this week so should hopefully be by the end of this, or early next week. I'll let you know when we have a date 👍

File: 12/umbraco-commerce/reference/storefront-api/endpoints/checkout.md
Comment (alina-tincas) [inline 12/umbraco-commerce/reference/storefront-api/endpoints/checkout.md:7]: Hi @mattbrailsford 👋 

We have a styleguide to follow to make the documentation more readable, where one of guides is to write [shorter sentences ](https://docs.umbraco.com/welcome/contribute/style-guide#long-sentences )with max 25 words. 

It seems in this case, these 2 sentences have more than 25 words. Can you please look into them? 😊 

_"The pay endpoint should then be launched in a WebView with this token which will then redirect to the given Orders payment gateway for payment capture." -_- my suggestion here would be to remove first "then"

To determine the outcome of the payment developers should monitor the WebViews URL which will be redirected to the same pay endpoint URL with either a `/completed`, `/canceled` or `/errored` suffix. - my suggestion here would be to split this into 2 sentences like so:
 "to determine the outcome of the payment developers should monitor the WebViews URL. Then they will be will be redirected to the same pay endpoint URL with either a `/completed`, `/canceled` or `/errored` suffix."

File: 12/umbraco-commerce/reference/storefront-api/endpoints/checkout.md
Comment (alina-tincas) [inline 12/umbraco-commerce/reference/storefront-api/endpoints/checkout.md:19]: these 2 sentences need to be shorten to 25 words as well 🙈 

1. To aid developers, the `initialize` endpoint will return the settings of the given Orders selected payment method, along with details of any expected metadata needed by the payment provider.
2. Developers can then use this data to perform an inline capture operation and then call the `confirm` endpoint when the capture is successful, passing back any metadata captured.

File: 12/umbraco-commerce/reference/storefront-api/endpoints/order.md
Comment (alina-tincas) [inline 12/umbraco-commerce/reference/storefront-api/endpoints/order.md:3]: In order to keep it to 25 words, this would be my suggestion:
```suggestion
The Order endpoints are where you will manage your carts/orders and perform cart management functionality. Some examples are adding items to the cart, or setting up billing and shipping details.
```

**Review by alina-tincas — CHANGES_REQUESTED** [review]
Hi @mattbrailsford 😊 

I have made a few suggestions that I was wondering you can take a look into? 🙌 

It is mostly about some [making shorter sentences](https://docs.umbraco.com/welcome/contribute/style-guide#long-sentences) with max 25 words. as per our guidelines to make the documentation more readable 🙈

File: 12/umbraco-commerce/reference/storefront-api/endpoints/order.md
Comment (mattbrailsford) [reply 12/umbraco-commerce/reference/storefront-api/endpoints/order.md]: Updated 👍

File: 12/umbraco-commerce/reference/storefront-api/endpoints/checkout.md
Comment (mattbrailsford) [reply 12/umbraco-commerce/reference/storefront-api/endpoints/checkout.md]: Updated 👍

File: 12/umbraco-commerce/reference/storefront-api/endpoints/checkout.md
Comment (mattbrailsford) [reply 12/umbraco-commerce/reference/storefront-api/endpoints/checkout.md]: Updated 👍

**Pull request merged**
